src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -416,6 +445,26 @@ export class MeetingManager implements AudioVideoObserver {
     this.publishSelectVideoInputDeviceError();
   };
 
+  invokeDeviceProvider(deviceLabels: DeviceLabels): void {
+    this.setupDeviceLabelTrigger(deviceLabels);
+
+    switch (true) {
+      case deviceLabels === DeviceLabels.AudioAndVideo || deviceLabels === DeviceLabels.Audio:
+        if (this.invokeAudioInputProvider) {
+          this.invokeAudioInputProvider();
+        }
+
+        if (this.invokeAudioOutputProvider) {
+          this.invokeAudioOutputProvider();
+        }
+
+      case deviceLabels === DeviceLabels.AudioAndVideo || deviceLabels === DeviceLabels.Video:
+        if (this.invokeVideoInputProvider) {
+          this.invokeVideoInputProvider();
+        }
```

Comment:
Is it good to write the code like this or better change to `if ... else ...` one?

Reply:
It's a bit opaque.

Reply:
This is to make the code concise. The basic logic here is, if `DeviceLabels` contains "Audio", we invoke audio input/output devices, if `DeviceLabels` contains "Video", we invoke video devices.

But we gonna change the invoke logic to `useEffect` one, so this could be unnecessary.

If we split the cases, the code is quite redundant.
```ts
  invokeDeviceProvider(deviceLabels: DeviceLabels): void {
    this.setupDeviceLabelTrigger(deviceLabels);

    switch (deviceLabels) {
      case DeviceLabels.Audio:
        if (this.invokeAudioInputProvider) {
          this.invokeAudioInputProvider();
        }
        if (this.invokeAudioOutputProvider) {
          this.invokeAudioOutputProvider();
        }
        break;
      case DeviceLabels.Video:
        if (this.invokeVideoInputProvider) {
          this.invokeVideoInputProvider();
        }
        break;
      case DeviceLabels.AudioAndVideo:
        if (this.invokeAudioInputProvider) {
          this.invokeAudioInputProvider();
        }

        if (this.invokeAudioOutputProvider) {
          this.invokeAudioOutputProvider();
        }

        if (this.invokeVideoInputProvider) {
          this.invokeVideoInputProvider();
        }
        break;
    }
  }
```

Reply:
Concise is less important than (a) clear, (b) easy to change safely, (c) easy to test. Your redundant case there is at least very easy to read and easy to tell that it's correct!

You could also try something like:

```typescript
// Audio input and output.
switch (deviceLabels) {
  case DeviceLabels.Audio:
  case DeviceLabels.AudioAndVideo:
    if (this.invokeAudioInputProvider) {
      this.invokeAudioInputProvider();
    }
    if (this.invokeAudioOutputProvider) {
      this.invokeAudioOutputProvider();
    }
    break;
  }
}

// Video input.
switch deviceLabels {
  case DeviceLabels.Video:
  case DeviceLabels.AudioAndVideo: {
    if (this.invokeVideoInputProvider) {
      this.invokeVideoInputProvider();
    }
    break;
  }
}
```

However, the existence of those conditional `invoke` things and this kind of complexity is a good indicator that some part of the design isn't right.

Reply:
Yes, you are right. I will keep this in mind. Thank you!